libs/core/include/core/bloom_filter.hpp

```diff
@@ -0,0 +1,224 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "core/bitvector.hpp"
+
+#include <cstddef>
+#include <functional>
+#include <iterator>
+#include <utility>
+#include <vector>
+
+namespace fetch {
+
+namespace byte_array {
+class ConstByteArray;
+}
+
+namespace internal {
+
+class HashSource;
+
+/*
+ * An ordered collection of hash functions for generating pseudorandom
+ * std::size_t indices for the Bloom filter. To apply the functions,
+ * to an input, invoke the factory's operator() and use the resulting
+ * HashSource.
+ *
+ * The factory must be kept alive while its HashSource instances remain
+ * in use.
+ */
+class HashSourceFactory
+{
+public:
+  using Bytes     = fetch::byte_array::ConstByteArray;
```

Comment:
I would prefer no to have this type alias because I will (at some point) remove byte array and its friends from the core namespace and leave it in the main fetch namespace. When this happens it will be an easy job to remove all the forwarding type aliases we have (provided they have the same name 😁 )

Reply:
It's gone.